src/index.ts

```diff
@@ -0,0 +1,207 @@
+import { BigNumber } from 'bignumber.js';
+
+export type ETHAddress = string;
+export type ENSName = string;
+export type SolodityType = string;
+export const EIP681NamedParameters = ['value', 'gas', 'gasLimit', 'gasPrice'];
+
+export type EIP681Object = {
+    scheme: 'ethereum';
+    prefix?: 'pay' | string;
+    /**
+     * Target Address in the format `0x1234DEADBEEF5678ABCD1234DEADBEEF5678ABCD` or `doge-to-the-moon.eth`
+     */
+    target_address: ETHAddress | ENSName;
+    /**
+     * The function the user is aiming to execute
+     */
+    function_name?: string;
+    /**
+     * The chain at which this action should be performed
+     * If undefined assume the current user's chain.
+     */
+    chain_id?: `${number}`;
+    /**
+     * Named variables
+     */
+    parameters?: Partial<{
+        value: `${number}`;
+        gas: `${number}`;
+        gasPrice: '${number}';
+    }>;
+    /**
+     * Function Arguments
+     * These arguments are fed to the function that is to be executed
+     */
+    args?: [SolodityType, string][];
+};
+
+const number_regex =
+    /^(?<major>[+-]?\d+)(?:\.(?<minor>\d+))?(?:[Ee](?<exponent>\d+))?$/;
+const prefix_regex = '(?<prefix>[a-zA-Z]+)-';
+const address_regex =
+    '(?:0x[\\w]{40})|(?:[a-zA-Z0-9][a-zA-Z0-9-]{1,61}[a-zA-Z0-9].[a-zA-Z]{2,})';
+
+// Full regex for matching
+const full_regex = `^ethereum:(?:${prefix_regex})?(?<address>${address_regex})\\@?(?<chain_id>[\\w]*)*\\/?(?<function_name>[\\w]*)*(?<query>\\?.*)?`;
+
+/**
+ * Cleanup any unresolved values in the query parameters
+ *
+ * Converts '2.014e18' to '2014000000000000000'
+ *
+ * @param {string} variable string.
+ * @param {string} value string.
+ *
+ * @return {string}
+ */
+function processValue(variable: string, value: string): string {
+    const isReserved = EIP681NamedParameters.includes(variable);
+    const isNumber = number_regex.test(value);
+
+    if (isReserved && !isNumber)
+        throw new Error(variable + ' needs to be a number');
+
+    if (isNumber) {
+        const match = value.match(number_regex).groups;
+
+        value = new BigNumber(
+            `${match.major}${match.minor ? '.' + match.minor : ''}${
+                match.exponent ? 'e+' + match.exponent : ''
+            }`,
+            10
+        ).toString();
+    }
+
+    return value;
+}
+
+/**
+ * Cleanup any unresolved values in the query parameters
+ *
+ * Converts '2014000000000000000' to '2.014e18'
+ *
+ * @param {string} variable string.
+ * @param {string} value string.
+ *
+ * @return {string}
+ */
+function stringifyValue(variable: string, value: string): string {
+    const isNumber = value.match(/^\d+$/); // Should really be the javascript version but ¯\_(ツ)_/¯ for now
```

Comment:
there are other ways of doing this that are more readable like  !isNaN

Reply:
From what it appears isNaN will also trigger for `0x`. I could use !isNaN and then override it for things starting with `0x`, or we could have a proper regex or some other way... suggestions?